src/lib/4337.js

```diff
@@ -0,0 +1,249 @@
+    maxFeePerGas: ethers.BigNumber.from(userop.maxFeePerGas),
+    maxPriorityFeePerGas: ethers.BigNumber.from(userop.maxPriorityFeePerGas),
+    paymasterAndData: ethers.utils.arrayify(userop.paymasterAndData),
+    signature: ethers.utils.arrayify(userop.signature),
+  };
+
+  return newUserop;
+};
+
+const senderAccountExists = async (indexer, sender) => {
+  const url = `${indexer.url}/accounts/${sender}/exists`;
+
+  const resp = await fetch(url);
+
+  return resp.status === 200;
+};
+
+const generateUserOp = (
+  signerAddress,
+  sender,
+  senderAccountExists = false,
+  accountFactoryAddress,
+  callData
+) => {
+  const userop = {
+    sender,
+    nonce: ethers.BigNumber.from("0"),
+    initCode: ethers.utils.arrayify("0x"),
+    callData: ethers.utils.arrayify("0x"),
+    callGasLimit: ethers.BigNumber.from("0"),
+    verificationGasLimit: ethers.BigNumber.from("1500000"),
+    preVerificationGas: ethers.BigNumber.from("21000"),
+    maxFeePerGas: ethers.BigNumber.from("0"),
+    maxPriorityFeePerGas: ethers.BigNumber.from("1000000000"),
+    paymasterAndData: ethers.utils.arrayify("0x"),
+    signature: ethers.utils.arrayify("0x"),
+  };
+
+  // initCode
+  if (!senderAccountExists) {
+    const accountCreationCode = accountFactoryInterface.encodeFunctionData(
+      "createAccount",
+      [signerAddress, ethers.BigNumber.from(0)]
+    );
+
+    userop.initCode = ethers.utils.hexConcat([
+      accountFactoryAddress,
+      accountCreationCode,
+    ]);
+  }
+
+  // callData
+  userop.callData = callData;
+
+  return userop;
+};
+
+const prepareUserOp = async (
+  owner,
+  sender,
+  callData,
+  { indexer, erc4337, token }
+) => {
+  if (erc4337 === undefined || token === undefined) {
+    throw new Error("Invalid config object");
+  }
+
+  const accountFactoryAddress = erc4337.account_factory_address;
+
+  // check that the sender's account exists
+  const exists = senderAccountExists(indexer, sender);
+
+  // generate a userop
+  const userop = generateUserOp(
+    owner,
+    sender,
+    exists,
+    accountFactoryAddress,
+    callData
+  );
+
+  return userop;
+};
+
+const paymasterSignUserOp = async (erc4337, userop) => {
+  const rpcUrl = `${erc4337.rpc_url}/${erc4337.paymaster_address}`;
+
+  const body = {
+    jsonrpc: "2.0",
+    id: 1,
+    method: "pm_ooSponsorUserOperation",
+    params: [
+      userOpToJson(userop),
+      erc4337.entrypoint_address,
+      { type: erc4337.paymaster_type },
+      1,
+    ],
+  };
+
+  const resp = await fetch(rpcUrl, {
+    method: "POST",
+    body: JSON.stringify(body),
+  });
+
+  if (resp.status !== 200) throw new Error(response.error.message);
+
+  const response = await resp.json();
+
+  if (!response?.result?.length) {
+    throw new Error("Invalid response");
+  }
+
+  return userOpFromJson(response.result[0]);
+};
+
+const submitUserOp = async (erc4337, userop) => {
+  const rpcUrl = `${erc4337.rpc_url}/${erc4337.paymaster_address}`;
```

Comment:
so we first submit it to get it sign, then submit again? Cannot do that in one go (sign and submit)?

Reply:
First the paymaster needs to sign, then we need to sign the userop that includes the paymaster's signature. Otherwise someone could swap out the paymaster's signature without your approval.

It's the standard way to do it.